File: src/test/java/kitchenpos/menus/application/domain/tobe/MenuTest.java
package kitchenpos.menus.application.domain.tobe;

import kitchenpos.menus.tobe.domain.Menu;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertThrows;

/*
    - 메뉴에 속한 상품의 수량은 0 이상이어야 한다.
    - 메뉴의 가격이 올바르지 않으면 등록할 수 없다.
      - 메뉴의 가격은 0원 이상이어야 한다.
    - 메뉴에 속한 상품 금액의 합은 메뉴의 가격보다 크거나 같아야 한다.
    - 메뉴는 특정 메뉴 그룹에 속해야 한다.
    - 메뉴의 이름이 올바르지 않으면 등록할 수 없다.
      - 메뉴의 이름에는 비속어가 포함될 수 없다.
 */
public class MenuTest {

    @Test
    void 메뉴_생성() {
        final String name = "치킨";
        final int price = 20000;
        final Long menuGroupId = 1L;

        assertDoesNotThrow(() -> new Menu(name, price, menuGroupId));
    }

    @Test
    void 메뉴의_가격은_0원_이상() {
        final String name = "치킨";
        final int price = -20000;
        final Long menuGroupId = 1L;

        assertThrows(IllegalArgumentException.class, () -> new Menu(name, price, menuGroupId));
    }
}

File: src/test/java/kitchenpos/menus/application/tobe/MenuTest.kt
package kitchenpos.menus.application.tobe

import kitchenpos.common.price
import kitchenpos.fixture.MenuFixtures.menu
import kitchenpos.fixture.MenuFixtures.menuProduct
import kitchenpos.menus.tobe.domain.MenuProducts
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.math.BigDecimal


class MenuTest {
    @Test
    fun `메뉴 상품 가격은 상품 가격과 동기화를 위해 변경이 가능하다`() {
        //given
        val menuProduct = menuProduct(price = BigDecimal.valueOf(10000).price())
        val menu = menu(
            displayStatus = false,
            price = BigDecimal.valueOf(8000).price(),
            menuProducts = MenuProducts(listOf(menuProduct))
        )

        //when
        menu.changeMenuProduct(menuProduct.productId, BigDecimal.valueOf(5000).price())

        //then
        assertThat(menu.menuProducts.totalPrice).isEqualTo(BigDecimal.valueOf(5000).price())
    }

    @Test
    fun `전시 중인 메뉴의 가격이 메뉴상품의 총 가격이 보다 클 경우 전시 비활성 상태로 바뀐다`() {
        //given
        val menuProduct = menuProduct(price = BigDecimal.valueOf(10000).price())
        val menu = menu(
            displayStatus = true,
            price = BigDecimal.valueOf(8000).price(),
            menuProducts = MenuProducts(listOf(menuProduct))
        )

        //when
        menu.changeMenuProduct(menuProduct.productId, BigDecimal.valueOf(5000).price())

        //then
        assertThat(menu.menuProducts.totalPrice).isEqualTo(BigDecimal.valueOf(5000).price())
        assertThat(menu.displayStatus).isFalse()
    }
}
